source/library/Argo/Codec.hs

```diff
@@ -0,0 +1,280 @@
+module Argo.Codec where
+
+import Control.Applicative ((<|>))
+
+import qualified Argo.Json.Array as Array
+import qualified Argo.Json.Boolean as Boolean
+import qualified Argo.Json.Member as Member
+import qualified Argo.Json.Name as Name
+import qualified Argo.Json.Null as Null
+import qualified Argo.Json.Number as Number
+import qualified Argo.Json.Object as Object
+import qualified Argo.Json.String as String
+import qualified Argo.Json.Value as Value
+import qualified Argo.Result as Result
+import qualified Argo.Vendor.Transformers as Trans
+import qualified Control.Applicative as Applicative
+import qualified Control.Monad as Monad
+import qualified Data.Functor.Identity as Identity
+import qualified Data.Text as Text
+
+decodeWith :: ValueCodec a -> Value.Value -> Result.Result a
+decodeWith c = either fail pure
+    . Identity.runIdentity
+    . Trans.runExceptT
+    . Trans.runReaderT (decode c)
+
+encodeWith :: ValueCodec a -> a -> Value.Value
+encodeWith c x = snd
+    . Identity.runIdentity
+    . Trans.runStateT (encode c x)
+    . Value.Null
+    $ Null.fromUnit ()
+
+project :: (i -> f) -> CodecOf r w f o -> CodecOf r w i o
+project f c = c { encode = encode c . f }
+
+data CodecOf r w i o = Codec
+    { decode :: r o
+    , encode :: i -> w o
+    }
+
+instance (Functor r, Functor w) => Functor (CodecOf r w i) where
+    fmap f c = Codec
+        { decode = fmap f $ decode c
+        , encode = fmap f . encode c
+        }
+
+instance (Applicative r, Applicative w) => Applicative (CodecOf r w i) where
+    pure x = Codec
+        { decode = pure x
+        , encode = const $ pure x
+        }
+    cf <*> cx = Codec
+        { decode = decode cf <*> decode cx
+        , encode = \ i -> encode cf i <*> encode cx i
+        }
+
+instance (Applicative.Alternative r, Applicative.Alternative w) => Applicative.Alternative (CodecOf r w i) where
```

Comment:
This instance can't be used in practice:

``` hs
bad :: ValueCodec a
bad = Applicative.empty
```


```
• No instance for (Monad.MonadPlus Identity.Identity)
    arising from a use of ‘Applicative.empty’
• In the expression: Applicative.empty
  In an equation for ‘bad’: bad = Applicative.empty
```

Reply:
And the crux of the problem is the `encode` bit. For some `c :: ValueCodec a`, it has `encode c :: State Value a`, and `StateT s m a` only has an `Alternative` instance when the inner monad `m` has a `MonadPlus` instance. Since obviously `Identity` can't implement choice, `State` can't implement `Alternative`. 

I'm not sure, but I think this is related to https://github.com/kowainik/tomland/pull/316.